### PR TITLE
Fixes #13913 - split importing of host and facts

### PR DESCRIPTION
**Pull request by ares**

### CODE REVIEW

**Comment by ares** [comment]
it would be super awesome if this could get into 1.11 as well, so we'd mark it deprecated in there, it is very small and safe change

File: app/models/host/base.rb [hunk 1]
```diff
@@ -134,6 +134,8 @@ def import_facts(facts)
       # we are not importing facts for hosts in build state (e.g. waiting for a re-installation)
       raise ::Foreman::Exception.new('Host is pending for Build') if build?
 
+      facts[:domain].downcase!
```
Comment (lzap) [inline app/models/host/base.rb]: `undefined method downcase! for nil:NilClass` when discovering a host. This breaks Discovery plugin. I alredy finished testing 5.0 with 1.11 so I would rather not see it in the release.

Comment (lzap) [inline app/models/host/base.rb]: Please only downcase if it's not nil since during discovery phase, domain is not yet present. Edit: Use `try(:downcase)` as below.

File: app/models/host/managed.rb [hunk 2]
```diff
@@ -434,15 +434,13 @@ def info
     info_hash
   end
 
-  # JSON is auto-parsed by the API, so these should be in the right format
-  def self.import_host_and_facts(hostname, facts, certname = nil, proxy_id = nil)
+  def self.import_host(hostname, import_type, certname = nil, proxy_id = nil)
```
Comment (lzap) [inline app/models/host/managed.rb:442]: Can you add default option to `import_type = "puppet"` instead of explicitly assigning it?

File: app/models/host/managed.rb [hunk 3]
```diff
@@ -452,13 +450,22 @@ def self.import_host_and_facts(hostname, facts, certname = nil, proxy_id = nil)
...
+
     # if proxy authentication is enabled and we have no puppet proxy set and the upload came from puppet,
     # use it as puppet proxy.
-    if facts['_type'].blank? || facts['_type'] == 'puppet'
+    if import_type == 'puppet'
       host.puppet_proxy_id ||= proxy_id
     end
 
-    host.save(:validate => false) if host.new_record?
+    host
```
Comment (lzap) [inline app/models/host/managed.rb:461]: This is a contract change, why?

File: app/models/host/managed.rb [hunk 4]
```diff
@@ -452,13 +450,22 @@ def self.import_host_and_facts(hostname, facts, certname = nil, proxy_id = nil)
...
 
+    # if proxy authentication is enabled and we have no puppet proxy set, use it.
+    host.puppet_proxy_id ||= proxy_id
+
     # if proxy authentication is enabled and we have no puppet proxy set and the upload came from puppet,
     # use it as puppet proxy.
-    if facts['_type'].blank? || facts['_type'] == 'puppet'
+    if import_type == 'puppet'
       host.puppet_proxy_id ||= proxy_id
     end
```
Comment (lzap) [inline app/models/host/managed.rb]: Ditto. This changes behavior while the ticket looks like pure refactoring.

**Comment by lzap** [comment]
Tests failed. Can you also add an extra test that imports facts and host from minimum possible set of input facts?

File: app/models/host/managed.rb (hunk 2 above)
Comment (ares) [inline app/models/host/managed.rb:442]: I wanted to avoid it, you should always specify what you're trying to import. Take it as a step towards the state where puppet is just another cfgmgmt tool we support and it can be extracted to a plugin.

File: app/models/host/managed.rb (hunk 4 above)
Comment (ares) [inline app/models/host/managed.rb]: how does it change behavior? if type is puppet then associate a proxy, the only change is that we don't get all facts in this method but we use import_type directly, I don't see why it's not refactoring

File: app/models/host/managed.rb (hunk 3 above)
Comment (ares) [inline app/models/host/managed.rb:461]: So that you can use this method when you want to trigger import_facts later on it's result, also I don't think the save is needed at this point, especially with validation skipping. If we'd keep same contract, the method would return only true or nil, so you couldn't get to newly built host. Maybe a small example will clarify it better

``` ruby
h = Host.import_host(...) # class method, similar to Host.new
h.import_facts(...) # now we can call other methods on new instance
```

I'd say that important thing is that contract of import_host_and_facts hasn't changed, not that this is a new method, and git only shows it as part of old import_host_and_facts since the body is similar

File: app/models/host/base.rb (hunk 1 above)
Comment (ares) [inline app/models/host/base.rb]: good catch, will add test for this which should prevent breaking in future as well

**Comment by ares** [comment]
> Can you also add an extra test that imports facts and host from minimum possible set of input facts?

What do you mean by minimal set? I'm not aware of any required fact, parsing only set associations if a particular fact was imported. Since we skip validations on importing (:disappointed:) we can save host just with it's name I think.

File: app/models/host/managed.rb (hunk 3 above)
Comment (lzap) [inline app/models/host/managed.rb:461]: Ok that's valid argument.

File: app/models/host/base.rb [hunk 5]
```diff
@@ -134,6 +134,8 @@ def import_facts(facts)
       # we are not importing facts for hosts in build state (e.g. waiting for a re-installation)
       raise ::Foreman::Exception.new('Host is pending for Build') if build?
 
+      facts[:domain].downcase! unless facts[:domain].nil?
```
Comment (lzap) [inline app/models/host/base.rb]: Hmmm `facts[:domain].try(:downcase!)` is perhaps better.

**Comment by lzap** [comment]
I would like to see a test that runs both methods with minimum (edge) case. This would help to find the `downcase` on `nil` error. Discovery does use this interface in a different way with minimum input data.

**Comment by lzap** [comment]
> What do you mean by this? I'm not aware of any required fact, parsing only set associations if a particular fact was imported. Since we skip validations on importing (:disappointed:) we can save host just with it's name I think.

Additional note: By minimum set I mean really no facts, just what's
required to create new host (mac, ipaddress?)

## 

Later,
 Lukas #lzap Zapletal

File: app/models/host/base.rb (hunk 5 above)
Comment (ares) [inline app/models/host/base.rb]: I'd argue it's not :-) I'd like to avoid using #try, but if you insist...

**Comment by ares** [comment]
> By minimum set I mean really no facts, just what's required to create new host (mac, ipaddress?)

That's the thing, nothing is required for importing host

this will work

```
Host.import_host('name', 'type').save(:validate => true) # => true
```

but I can add at least this and the minimum set for import_facts (OS info is required I think)

**Comment by lzap** [comment]
Ok that looks like totally valid test for an edge case, please add it. I have no other comments, will test then.

**Comment by lzap** [comment]
FYI tests failed.

**Comment by ares** [comment]
[test] I believe that's false positive, @lzap could you rereview? there were more refactoring inside the import_host method as well

**Comment by lzap** [comment]
Ok LGTM, it works with Discovery develop and 5.0.1 as well, therefore I have no objections against pulling it in the 1.11 release. It will only introduce those deprecation warnings in logs. I am making the necessary changes for discovery to split that as well now. I will merge once Jenkins is happy (depends on https://github.com/theforeman/foreman/pull/3199)

**Comment by ares** [comment]
[test] @lzap would you mind changing PR flags?

**Comment by lzap** [comment]
Done. Please do not re-test until the mentioned PR fixes it.

**Comment by lzap** [comment]
Merged as d4ec64418b8db6093375f8d4b690d506705163f7, thank you!
